contracts/Migrator.sol

```diff
@@ -0,0 +1,239 @@
+    }
+
+    /**
+     * @notice Migrates the specified amount of OGV to OGN
+     * @param ogvAmount Amount of OGV to migrate
+     * @return ognReceived OGN Received
+     */
+    function migrate(uint256 ogvAmount) external isSolvent returns (uint256 ognReceived) {
+        return _migrate(ogvAmount, msg.sender);
+    }
+
+    /**
+     * @notice Migrates OGV stakes to OGN. Can also include unstaked OGN & OGV
+     *          balances from the user's wallet (if specified).
+     * @param lockupIds OGV Lockup IDs to be migrated
+     * @param ogvAmountFromWallet Extra OGV balance from user's wallet to migrate & stake
+     * @param ognAmountFromWallet Extra OGN balance from user's wallet to stake
+     * @param migrateRewards If true, Migrate & Stake received rewards
+     * @param newStakeAmount Max amount of OGN (from wallet+unstake) to stake
+     * @param newStakeDuration Duration of the new stake
+     */
+    function migrate(
+        uint256[] calldata lockupIds,
+        uint256 ogvAmountFromWallet,
+        uint256 ognAmountFromWallet,
+        bool migrateRewards,
+        uint256 newStakeAmount,
+        uint256 newStakeDuration
+    ) external isSolvent {
+        if (!isMigrationActive()) {
+            revert MigrationIsInactive();
+        }
+
+        if (lockupIds.length == 0) {
+            revert LockupIdsRequired();
+        }
+
+        // Unstake
+        (uint256 ogvAmountUnlocked, uint256 rewardsCollected) = ogvStaking.unstakeFrom(msg.sender, lockupIds);
+
+        if (migrateRewards) {
+            // Include rewards if needed
+            ogvAmountFromWallet += rewardsCollected;
+        }
+
+        ogvAmountFromWallet += ogvAmountUnlocked;
+
+        if (ognAmountFromWallet > 0) {
+            // Transfer in additional OGN to stake from user's wallet
+            ogn.transferFrom(msg.sender, address(this), ognAmountFromWallet);
+        }
+
+        // Migrate OGV to OGN and include that along with existing balance
+        ognAmountFromWallet += _migrate(ogvAmountFromWallet, address(this));
+
+        if (ognAmountFromWallet < newStakeAmount) {
+            revert InvalidStakeAmount();
+        }
+
+        uint256 ognToWallet = ognAmountFromWallet - newStakeAmount;
+        if (ognToWallet > 0) {
+            ogn.transfer(msg.sender, ognToWallet);
+        }
+
+        if (newStakeAmount > 0) {
+            // Stake it
+            ognStaking.stake(
+                newStakeAmount,
+                newStakeDuration,
+                msg.sender,
+                false,
+                -1 // New stake
+            );
+        }
+
+        emit LockupsMigrated(msg.sender, lockupIds, newStakeAmount, newStakeDuration);
+    }
+
+    /**
+     * @notice Migrates caller's OGV to OGN and sends it to the `receiver`
+     * @return ognReceived OGN Received
+     */
+    function _migrate(uint256 ogvAmount, address receiver) internal returns (uint256 ognReceived) {
+        if (!isMigrationActive()) {
```

Comment:
We check migration active both in an external method and again in this internal method. This means that a migrate with lockups checks this twice. Usually it's better to have a check happen purely at one layer, either internally only, of externally only.

However, I think we would be okay to drop the isMigrationActive checks from the migration methods. This would mean that migrations could continue until we decommission, but that should be fine. Will make for less gas during migrations.

We decommission, the OGN leaving the contract will stop migrations.

It's fine if we allow coin migrations before the official start. staking migrations won't work, of course until start has been called.

Reply:
That makes sense. Have made the change. However do note that people can start using the Migrator contract as soon as it's deployed and funded.

So, let's say we send 10M OGN to the contract for migrations but we only need 9M or so (and haven't called `transferExcessTokens` yet), There's a chance that the `isSolvent` won't catch if for some reason people get more OGN than expected for that additional 1M OGN

Reply:
Good points.

We'll just have to keep testing amounts in it until we are ready to go, and perhaps our go live action will pull any excess from it just before/after the start().